src/core/model_concrete.jl

```diff
@@ -0,0 +1,257 @@
+export ABM, StandardABM, UnkillableABM, FixedMassABM
+using StaticArraysCore: SizedVector
+
+ContainerType{A} = Union{AbstractDict{Int,A}, AbstractVector{A}}
+
+# And the three implementations here are just variants with different `C` type.
+struct SingleContainerABM{S<:SpaceType,A<:AbstractAgent,C<:ContainerType{A},F,P,R<:AbstractRNG} <: AgentBasedModel{S,A}
+    agents::C
+    space::S
+    scheduler::F
+    properties::P
+    rng::R
+    maxid::Base.RefValue{Int64}
+end
+
+const SCABM = SingleContainerABM
+const StandardABM{A,S} = SingleContainerABM{A,S,Dict{Int,A}}
+const UnkillableABM{A,S} = SingleContainerABM{A,S,Vector{A}}
+const FixedMassABM{A,S} = SingleContainerABM{A,S,SizedVector{A}}
+
+containertype(::SingleContainerABM{S,A,C}) where {S,A,C} = C
+
+function construct_agent_container(container, A)
+    if container <: Dict
+        return Dict{Int,A}
+    elseif container <: Vector
+        return Vector{A}
+    else
+        throw(ArgumentError("Unrecognised container $container, please specify either Dict or Vector."))
+    end
+end
+
+"""
+    union_types(U::Type)
+Return a tuple of types within a `Union`.
+"""
+union_types(T::Type) = (T,)
+union_types(T::Union) = (union_types(T.a)..., union_types(T.b)...)
+
+"""
+    SingleContainerABM(AgentType [, space]; properties, kwargs...) → model
+
+A concrete version of [`AgentBasedModel`](@ref) that stores all agents in a
+single container. Offers the variants:
+
+- [`StandardABM`](@ref)
+- [`UnkillableABM`](@ref)
+- [`FixedMassABM`](@ref)
+"""
+function SingleContainerABM(
+    ::Type{A},
+    space::S = nothing;
+    container::Type = Dict{Int,A},
+    scheduler::F = Schedulers.fastest,
+    properties::P = nothing,
+    rng::R = Random.default_rng(),
+    warn = true
+) where {A<:AbstractAgent,S<:SpaceType,F,P,R<:AbstractRNG}
+    agent_validator(A, space, warn)
+    C = construct_agent_container(container, A)
+    agents = C()
+    return SingleContainerABM{S,A,C,F,P,R}(agents, space, scheduler, properties, rng, Ref(0))
+end
+
+function SingleContainerABM(agent::AbstractAgent, args...; kwargs...)
+    return SingleContainerABM(typeof(agent), args...; kwargs...)
+end
+
+
+"""
+    StandardABM(AgentType [, space]; properties, kwargs...) → model
+
+
+The most standard concrete implementation of an [`AgentBasedModel`](@ref),
+as well as the default version of the generic [`AgentBasedModel`](@ref) constructor.
+`StandardABM` stores agents in a dictionary mapping unique `Int` IDs to agents.
+See also [`UnkillableABM`](@ref), [`FixedMassABM`](@ref).
+"""
+StandardABM(args...; kwargs...) = SingleContainerABM(args...; container=Dict)
```

Comment:
Super minor suggestion:
```suggestion
StandardABM(args...; kwargs...) = SingleContainerABM(args...; container=Dict{Int})
```